### PR TITLE
WIP: implementation

### PR DESCRIPTION
**Pull request by i4ki**
Implements the bitwise calculator.

The language is limited to assignment and expressions. Eg.:

```c
bwc> a = 0b10101010
bin: 10101010
hex: aa
bwc> a&0xff
bin: 10101010
hex: aa
bwc> a|0xff
bin: 11111111
hex: ff
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+func (_ Int) Type() Nodetype  { return NodeInt }
+func (i Int) String() string { return strconv.Itoa(int(i)) }
+
+func (_ Expr) Type() Nodetype { return NodeExpr }
+func (a Expr) String() string {
+	lhs := a.Lhs.String()
+	rhs := a.Rhs.String()
+	op := a.Op.String()
+
+	return fmt.Sprintf("%s %s %s", lhs, rhs, op)
```
Comment (katcipis) [inline main.go]: There seems to be some duplication between this and the **desc** function on the tests (perhaps for good reason, just pointing ou)

File: main.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	case OpAND:
+		ret = lhs & rhs
+	case OpOR:
+		ret = lhs | rhs
+	}
+	return ret
+}
+
+func main() {
+	expr := Expr{
```
Comment (katcipis) [inline main.go]: Someone is on the mood for some junction/disjunction action ? =P

Comment (katcipis) [reply main.go]: Read the README later (stupid) now I got the idea of bitwise calculator, pretty cool =D, I'm considerably stupid for this kind of thing too =)

**Comment by i4ki** [comment]
Sorry, for the 'out of context code'. Yesterday, after several mistakes writing and reading some bitwise operations in osdev, I had an idea of integrating a tool in acme to evaluate bitwises. You can use in two ways:

1. Typing "win bwc" to get an interactive calculator.
2. Select some bitwise in a code you're reviewing or writing and then middle click in bwc.

As I'm in good mood nowadays for everything, I'm gonna implement this soon before I get into darkness again =)

**Comment by ryukinix** [comment]
WoW, this is impressive! I want test this.

File: README.md [hunk 3]
```diff
@@ -2,18 +2,69 @@
 
 Because I'm dumb.
 
-# language
+# The tool
 
-Mathematical reverse polish notation language.
+The language was designed in a such a way that
+makes easy to copy-paste bitwises from code
+in most programming languages (at least from
```
Comment (katcipis) [inline README.md:9]: Perhaps "at least C descendents"

File: README.md [hunk 4]
```diff
@@ -2,18 +2,69 @@
...
 ```
+
+# The language
+
+```bnf
+letter		= "a".."z" | "A".."Z";
+digit		= "0".."9";
+alphanum	= letter | digit;
+
+decdigit 	= "0".."9";
```
Comment (katcipis) [inline README.md]: what is the difference between decdigit and digit ?

File: README.md [hunk 5]
```diff
@@ -2,18 +2,69 @@
...
+hexdigit 	= decdigit | "a".."f";
+bindigit 	= "0" | "1";
+
+decimal	= decdigit { decdigit };
+hexadecimal	= "0x" hexdigit { hexdigit };
+binary		= "0b" bindigit { bindigit };
+
+number		= decimal | hexadecimal | binary;
+ident		= alphanum {alphanum};
```
Comment (katcipis) [inline README.md]: hmm "7987" would be a valid identifier for a variable ? (nothing against it for now, just trying to understand =P)

File: README.md [hunk 6]
```diff
@@ -2,18 +2,69 @@
...
+binaryop	= "&" | "|" | "^" | "<<" | ">>";
+unaryop		= "~";
+expr		= [ "(" ] unaryexpr | binaryexpr [ ")" ];
+binaryexpr	= (expr | binaryexpr | unaryexpr | number) binaryop
```
Comment (katcipis) [inline README.md]: Perhaps:

```
binaryexpr_op = (expr | binaryexpr_op | unaryexpr | number)
binaryexpr	= binaryexpr_op binaryop binaryexpr_op
```

To avoid duplication ? Feel free to ignore the naming (or the whole idea =P)

File: infix/ast.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	case OpSHL:
+		return "<<"
+	case OpSHR:
+		return ">>"
+	}
+
+	return "<invalid op>"
+}
+
+func (_ Int) Type() Nodetype { return NodeInt }
```
Comment (katcipis) [inline infix/ast.go]: You don't need to write "_ Int", just "Int" is enough to ignore parameters. Example:

https://play.golang.org/p/C_zbybVB06M

File: infix/ast.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+func (i Int) String() string { return strconv.Itoa(int(i)) }
+
+func (_ BinExpr) Type() Nodetype { return NodeBinExpr }
+func (a BinExpr) String() string {
+	lhs := a.Lhs.String()
+	rhs := a.Rhs.String()
+	op := a.Op.String()
+
+	return fmt.Sprintf("%s %s %s", lhs, rhs, op)
```
Comment (katcipis) [inline infix/ast.go]: I think that just this line suffices (AFAIK Go automatically calls the String method of the type if there is one)

File: infix/ast.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func (_ UnaryExpr) Type() Nodetype { return NodeUnaryExpr }
+func (a UnaryExpr) String() string {
+	return fmt.Sprintf("%s%s",
+		a.Op.String(), a.Value.String())
```
Comment (katcipis) [inline infix/ast.go]: Same thing here, no need to call String explicitely

Comment (katcipis) [reply infix/ast.go]: Found the docs on this: https://tour.golang.org/methods/17

File: infix/eval.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+	if n.Type() == NodeUnaryExpr {
+		return evalUnaryExpr(n.(UnaryExpr))
+	}
+
+	if n.Type() != NodeBinExpr {
+		return 0, fmt.Errorf("unexpected %s", n)
+	}
+
+	return evalBinExpr(n.(BinExpr))
```
Comment (katcipis) [inline infix/eval.go]: This logic confused me a little, it seems more easy to understand if all conditionals check for a type and the default case returns the error (the last return), or use a switch case

File: infix/eval.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+func evalUnaryExpr(expr UnaryExpr) (Int, error) {
+	num, err := Eval(expr.Value)
+	if err != nil {
+		return 0, err
+	}
+
+	switch expr.Op {
+	case OpNOT:
+		return Int(^int64(num)), nil
```
Comment (katcipis) [inline infix/eval.go:58]: Just curious, is the cast needed here ?

File: infix/eval.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+	if expr.Lhs.Type() != NodeInt {
+		lhs, err = Eval(expr.Lhs)
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		lhs = expr.Lhs.(Int)
+	}
+
+	if expr.Rhs.Type() != NodeInt {
```
Comment (katcipis) [inline infix/eval.go]: hmmm the sweet smell of duplication =P. Perhaps some sort of "evalOperand" that does this if/else eval/casting logic ?

File: infix/eval.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	var ret Int
+	switch expr.Op {
+	case OpAND:
+		ret = lhs & rhs
+	case OpOR:
+		ret = lhs | rhs
+	case OpSHL:
+		ret = lhs << uint(rhs)
+	case OpSHR:
+		ret = lhs >> uint(rhs)
```
Comment (katcipis) [inline infix/eval.go:92]: If for some reason a new operator is added and no code is added here it will pass silently but will produce unexpected results (nothing)

Comment (katcipis) [reply infix/eval.go:92]: It seems interesting to have a default clause that fails just in case (and to make errors explicit)

File: infix/eval_test.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
+package infix
+
+import (
+	"fmt"
+	"testing"
+)
+
+var format = fmt.Sprintf
```
Comment (katcipis) [inline infix/eval_test.go:8]: lazy moderfucker =P

File: infix/eval_test.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+	} else if n.Type() == NodeUnaryExpr {
+		a := n.(UnaryExpr)
+		return format("%s%s", a.Op, a.Value)
+	}
+
+	a := n.(Int)
+	return format("%d", a)
+}
+
+func TestEval(t *testing.T) {
```
Comment (katcipis) [inline infix/eval_test.go]: On this case it is more easy to understand but on nash I had a real pain on understanding eval tests that are performed on nodes that are built by hand like this, it is much more hard to understand than something like:

```
expr: "0 & 0",
res: 0
}
```

But this would imply tests that covers the parsing too, together with eval. But they are still isolated on the sense that there is no source of non-determinism and avoid duplicating test coverage (you need to test the parsing too I think). So on nash I was pretty torn if it would not be better to always test on top of the parser/lexer instead of building the AST nodes by hand on the tests.

Another advantage are tests that are on top of the language specification, not the design of node types and the AST. But I'm not sure, just brainstorming ;-).

**Review by katcipis — CHANGES_REQUESTED** [review]
Just small remarks and some brainstorming =D. I'm not finished...you wrote a huge amount of code =P

File: infix/lexer.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+// The name isn't eoi to avoid confusion.
+const eof = -1
+
+func (t Tokval) String() string {
+	return fmt.Sprintf("Token(%s, %s)", t.T, t.V)
+}
+
+// Lex creates a concurrent lexer and returns a
+// channel of tokens processed from input.
+func Lex(input string) chan Tokval {
```
Comment (katcipis) [inline infix/lexer.go]: Can't we restrict the direction of the channel here ? (I suppose it could be read only)

File: infix/lexer.go [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+		input:  input,
+		tokens: make(chan Tokval),
+	}
+
+	go l.run()
+	return l.tokens
+}
+
+// run the state machine
+func (l *lexer) run() {
```
Comment (katcipis) [inline infix/lexer.go:50]: This is one thing that I was discussing on other projects with you, should core modules create goroutines internally or just expose functions that will generate some sort of loop and let the caller decide if it wants to create a goroutine for it or just block the main one with it (like the ListenAndServe).

The usage would be:

```go
l := Lex()
go func() {
    for token := range l.Tokens() {
    }
}()

l.Run()
```

The user can chooses what code runs on which goroutine and there is no obligatory goroutine creation by calling any function (the user is always on control). Not sure of what would be best, just thinking about this lately. I'm prone to not creating the goroutine because sometimes the user can use the main gorotine to run the loop, if all functions always creates goroutines internally the user has no choice and will have to lock the main goroutine on some kind of nil channel.

File: infix/lexer.go [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+	l.start = l.pos
+}
+
+// accept checks if next is in the valid string.
+func (l *lexer) accept(valid string) bool {
+	return strings.IndexRune(valid, l.peek()) != -1
+}
+
+// acceptRun consumes next runes if valid.
+func (l *lexer) acceptRun(valid string) {
```
Comment (katcipis) [inline infix/lexer.go:123]: Im having trouble trying to understand why the function that consumes runes is called "acceptRun" =/

File: infix/lexer.go [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+	default:
+		return l.errorf("Unexpected %q at %d", r, l.pos)
+	}
+	return nil
+}
+
+// lexer of dec, hex and bin numbers.
+func lexNumber(l *lexer) stateFn {
+	r := l.next()
+	next := l.peek()
```
Comment (katcipis) [inline infix/lexer.go:205]: Feel free to ignore because this may be just me being overzealous with small functions and etc =P, but I was thinking on something like:

```go
if isBinary(r, next) {
    handleBinary(l)
} else if isHexadecimal(r, next) {
    handleHexadecimal(l)
}
```

File: infix/lexer_test.go [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+	t.Helper()
+	got := consume(infix.Lex(tc.in))
+	if len(got) != len(tc.out) {
+		t.Logf("test data: %v", tc.in)
+		t.Logf("got: %v", got)
+		t.Fatalf("expect %d elems but got %d",
+			len(tc.out), len(got))
+	}
+
+	for i := 0; i < len(tc.out); i++ {
```
Comment (katcipis) [inline infix/lexer_test.go:31]: at least a "want vs got" could be used instead of e/g =P... you lazy moderfucker

File: infix/lexer_test.go [hunk 21]
```diff
@@ -0,0 +1,256 @@
...
+		if e.T != g.T {
+			t.Fatalf("tok differs: %v != %v", e, g)
+		}
+		if e.V != g.V {
+			t.Fatalf("tok differs: %s != %s", e.V, g.V)
+		}
+	}
+}
+
+func TestLexer(t *testing.T) {
```
Comment (katcipis) [inline infix/lexer_test.go:43]: We have some testing for errors parsing numbers but not here for operators, lets test those boundary conditions and corner cases =D

File: infix/parser.go [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+	tok, ok := <-p.tokens
+	if !ok {
+		return TokEOF
+	}
+	return tok
+}
+
+func (p *parser) parseExpr() (Node, error) {
+	var eoferr = func(expect string) error {
+		return fmt.Errorf("premature eof, expects %s",
```
Comment (katcipis) [inline infix/parser.go]: Perhaps "premature eof, expected %s" would be more clear (not sure, my english is far from great)

File: infix/parser.go [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+			expect)
+	}
+
+	tok := p.peek()
+	if tok.T == EOF {
+		return nil, eoferr("expression")
+	}
+
+	// left hand side of expr
```
Comment (katcipis) [inline infix/parser.go]: Instead of two comments "left hand side" and "right hand side" we could have two functions

File: infix/parser.go [hunk 24]
```diff
@@ -0,0 +1,224 @@
...
+		}
+
+		if intstr[1] == 'x' {
+			val, err := strconv.ParseInt(intstr[2:], 16, 64)
+			return Int(val), false, err
+		}
+	}
+
+	val, err := strconv.ParseInt(intstr, 10, 64)
+	return Int(val), false, err
```
Comment (katcipis) [inline infix/parser.go:254]: Will the error of strconv have enough context to understand the problem ?

File: infix/token.go [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+	case OR: return "|"
+	case AND: return "&"
+	case XOR: return "^"
+	case NOT: return "~"
+	case SHL: return "<<"
+	case SHR: return ">>"
+	case Illegal: return "<ileggal>"
+	case EOF: return "EOF"
+	}
+	panic("invalid token")
```
Comment (katcipis) [inline infix/token.go]: This makes me think on the need of the language to have some sort of enumeration that does not allow the type to exists with a invalid value (validated statically). For these cases the simplification of only having consts feels flat =/

File: main.go [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+	"os"
+	"strings"
+
+	"github.com/madlambda/bwc/infix"
+)
+
+func main() {
+	for {
+		fmt.Printf("bwc> ")
+		var line [256]byte
```
Comment (katcipis) [inline main.go]: Although I admit that a line that is bigger than 256 bytes is really hard to exist I dislike arbitrary limits like this. Can't we use the [bufio.Scanner](https://golang.org/pkg/bufio/#NewScanner) here ? I think the code would get smaller and we remove the arbitrary limit on the line length (or perhaps it is not arbitrary).

**Review by katcipis — CHANGES_REQUESTED** [review]
More crappy remarks =D

File: README.md (hunk 4 above)
Comment (i4ki) [reply README.md]: none, duplicate code =P

File: README.md (hunk 5 above)
Comment (i4ki) [reply README.md]: no, it's invalid. Thanks

File: infix/ast.go (hunk 7 above)
Comment (i4ki) [reply infix/ast.go]: I know, but I find this confuse. Dont you?

File: infix/ast.go (hunk 8 above)
Comment (i4ki) [reply infix/ast.go]: yes. This logic was more complicated in first implementation and then I failed to visualize this when refactored. Thanks.

File: infix/eval.go (hunk 11 above)
Comment (i4ki) [reply infix/eval.go:58]: it's not a cast but a conversion. As the return type is Int and the value int64 it's probably required.

File: infix/eval.go (hunk 10 above)
Comment (i4ki) [reply infix/eval.go]: I'll improve.

File: infix/eval.go (hunk 12 above)
Comment (i4ki) [reply infix/eval.go]: I was unable to name this thing. Thanks

File: infix/eval.go (hunk 13 above)
Comment (i4ki) [reply infix/eval.go:92]: Yes, a default clause is good to catch parser errors (in this case) as we do not validate every aspect of the ast.

File: infix/lexer.go (hunk 16 above)
Comment (i4ki) [reply infix/lexer.go]: What you need to avoid?

File: infix/eval_test.go (hunk 15 above)
Comment (i4ki) [reply infix/eval_test.go]: Agreed, but some tests on top of AST is good to ensure it's built correctly. Some familiarity with the ast is good also, to understand the evaluator.

There are missing tons of tests! I'll use code for next ones.

File: infix/ast.go (hunk 7 above)
Comment (katcipis) [reply infix/ast.go]: Not that much to be honest, make sense to me to be able to omit where you can do it safely and unambiguously, but I can see your point too and have nothing against "_" =)

File: infix/eval.go (hunk 11 above)
Comment (katcipis) [reply infix/eval.go:58]: I see, but anyway wont the "^" operator work directly on num ? like "^num", does it have to be ^int64(num) ? (I'm curious on the behavior of operators on named types but I'm lazy to search the docs =P)

File: infix/eval.go (hunk 12 above)
Comment (katcipis) [reply infix/eval.go]: naming is hard x_x. I find amusing when sometimes on the next day I look at something and the name is obvious but on the previous day it seemed impossible to name it, our minds are pretty bizarre on how they work =P

File: infix/lexer.go (hunk 16 above)
Comment (katcipis) [reply infix/lexer.go]: The caller writing things or closing the channel, since both things seems to go against the original intent of your API. I like APIs to have the smallest surface possible and to be explicit. Leaving as it is right now you are saying to me (with code) that I can write and close this channel (you can use docs to say that I should not, but if you can say it with code I prefer the code to do the talking =P). But this may be just me...being pedantic and annoying =D

File: infix/lexer.go (hunk 17 above)
Comment (i4ki) [reply infix/lexer.go:50]: > I'm prone to not creating the goroutine because sometimes the user can use the main gorotine to run the loop,

But what's the advantages? This way is simple and the use is very straightforward:

```go
for tok := range Lex(src) {
    fmt.Printf("token: %s\n", tok)
}
```
The other way requires documentation to make clear that calling Run() and Tokens() on same goroutine deadlocks. I understand the benefits on more complex tasks but fail to see any point here.

> if all functions always creates goroutines internally the user has no choice and will have to lock the main goroutine on some kind of nil channel.

I dont get this. Only the producer ran on a new goroutine, the consumer can run anywhere you want.

File: infix/lexer.go (hunk 18 above)
Comment (i4ki) [reply infix/lexer.go:123]: because it accepts a run of runes =P 
A run of runes in the sense of a series of runes. I can rename if you want.

File: infix/lexer.go (hunk 19 above)
Comment (i4ki) [reply infix/lexer.go:205]: This function is called 'lexNumber' and binary/hex is only handled here.
If there existed other places where handleBinary or handleHex was needed, I'm sure I would have put it on separate functions.

File: infix/lexer_test.go (hunk 20 above)
Comment (i4ki) [reply infix/lexer_test.go:31]: got was already taken =P

File: infix/lexer_test.go (hunk 21 above)
Comment (i4ki) [reply infix/lexer_test.go:43]: Do you mean test numbers close to operators?

File: infix/parser.go (hunk 22 above)
Comment (i4ki) [reply infix/parser.go]: hmm Maybe expected is better, indeed.

File: infix/lexer.go (hunk 17 above)
Comment (katcipis) [reply infix/lexer.go:50]: I was generalizing an idea, but I think this is not the best example to discuss it =P. Just forget about it

File: infix/lexer.go (hunk 18 above)
Comment (katcipis) [reply infix/lexer.go:123]: Never heard of the term "run of runes", perhaps that is why I did not understood it =)

File: infix/parser.go (hunk 23 above)
Comment (i4ki) [reply infix/parser.go]: yes, I commonly implement this way first to visualize *what* makes sense to extract in function and *how*. My brain is very limited.
Will improve.

File: infix/parser.go (hunk 24 above)
Comment (i4ki) [reply infix/parser.go:254]: probably not =P Error messages are pretty dumb for now.

File: infix/lexer.go (hunk 19 above)
Comment (katcipis) [reply infix/lexer.go]: I like to use functions to avoid having to write comments. that seems to be the case here (not just because the function is going to be used on more than one place). Having to comment what a function is doing inside it is usually a hint that I could extract a function to say that, specially if the overhead is negligible.

File: infix/token.go (hunk 25 above)
Comment (i4ki) [reply infix/token.go]: yes, is very painful to handle that ourselves every time.

File: main.go (hunk 26 above)
Comment (i4ki) [reply main.go]: hahaha sorry for that
This main file was just my 'test binary' to play in the acme win. I dont like bufio.Scanner because it's a little monster and hard to configure.. then I made something fast to use. 
In our case, the statements could span multiple lines, then we will need to continue to read lines from input while parser notifies/reports 'premature eof' error (maybe using error types or err behaviours).

**Comment by i4ki** [comment]
@katcipis Thanks for the detailed review! Helped a lot.

File: infix/eval.go (hunk 11 above)
Comment (i4ki) [reply infix/eval.go:58]: hmm good question. I'm lazy too =) hahaha

File: infix/lexer.go (hunk 16 above)
Comment (i4ki) [reply infix/lexer.go]: sorry, I had leave this to reply later and forgot. 
Nice, just to know =)
Commonly I dont add channel direction when the channel is not exposed, then... sometimes I forget it in APIs.
There's a controversial go issue proposing the drop of channel direction in Go2 by Griesemer and I share their concerns. As far as I remember, Rob Pike, Bradfitz and others were very against this proposal. But as channels returning from APIs are not good in lots of cases *and* channel direction is very rarely used for internal code, dropping the feature would be nice (a way to fight against code complexity/inflation).

But I like simple APIs returning channels also, then... I'm not sure yet if makes sense drop this or not.

Thanks!

File: infix/lexer.go (hunk 18 above)
Comment (i4ki) [reply infix/lexer.go:123]: got this naming from Rob Pike presentation on writing lexers. The name did makes sense at the time probably because I often see it being used in math, as in _a run of consecutive numbers_.

Comment (ryukinix) [reply infix/lexer.go:123]: Hm, but a **run** itself is a consecutive sequence of numbers. This is your [definition](http://www.maths.surrey.ac.uk/hosted-sites/R.Knott/runsums/index.html):

>  A run of numbers is a sequence of consecutive whole numbers i.e. with no gaps in the sequence:  such as [2, 3, 4], but not [5, 6, 8, 9] (because 7 is missing) neither [2, 3, 3] (repeated 3)

I don't see the mathematical `run` making sense with runes, except if run was redefined by someone like Rob Pike does and now it have a more broad definition.

Wouldn't better just use a more generic concept and common in programming? Like sequence, I don't know. `matchSeq`or something like that.
